# Session 15 - Automated EDA Tools & Comparative Analysis

Comprehensive solutions for automated Exploratory Data Analysis (EDA) tasks using **Pandas Profiling**, **Sweetviz**, and **D-Tale**.

### Task 1: Spotify Top 100 Songs - Profiling & Data Quality Analysis

**Objective**: Generate a profile report for Spotify Top 100 Songs dataset, inspect missing values, and identify data type issues.

In [1]:
import pandas as pd
import numpy as np

# Load Spotify Top 100 Songs dataset
spotify_df = pd.read_csv('session15_data/spotify_top100.csv')

# Dataset Structure & Data Types
print('=== Dataset Summary ===')
spotify_df.info()

# Check Missing Values
print('\n=== Missing Values Count ===')
print(spotify_df.isnull().sum())

# Inspect Data Type Issues
print('\n=== Inspecting Popularity Column ===')
print('Sample values:', spotify_df['Popularity'].head(5).tolist())
print('Data Type:', spotify_df['Popularity'].dtype)

# Standard ydata-profiling code syntax:
# from ydata_profiling import ProfileReport
# profile = ProfileReport(spotify_df, title='Spotify Top 100 Profiling Report')
# profile.to_file('session15_data/spotify_profile_report.html')


=== Dataset Summary ===
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          100 non-null    int64  
 1   Title         100 non-null    str    
 2   Artist        100 non-null    str    
 3   Genre         96 non-null     str    
 4   BPM           100 non-null    int64  
 5   Energy        100 non-null    int64  
 6   Danceability  100 non-null    int64  
 7   Loudness      100 non-null    int64  
 8   Liveness      100 non-null    int64  
 9   Valence       100 non-null    int64  
 10  Length        100 non-null    int64  
 11  Acousticness  90 non-null     float64
 12  Popularity    100 non-null    str    
dtypes: float64(1), int64(8), str(4)
memory usage: 12.7 KB

=== Missing Values Count ===
Rank             0
Title            0
Artist           0
Genre            4
BPM              0
Energy           0
Danceability     0
Loudness        

### Task 2: Zomato Restaurants - Mumbai vs Delhi Comparison (Sweetviz)

**Objective**: Create a visual comparison report between Zomato restaurant data for Mumbai and Delhi using `sweetviz.compare()`.

In [2]:
import pandas as pd
import sweetviz as sv

# Load datasets
df_mumbai = pd.read_csv('session15_data/zomato_mumbai.csv')
df_delhi = pd.read_csv('session15_data/zomato_delhi.csv')

# Generate Sweetviz comparison report
report = sv.compare([df_mumbai, 'Mumbai'], [df_delhi, 'Delhi'])
report.show_html('session15_data/zomato_sweetviz_compare.html', open_browser=False)

print('Comparison Report saved to: session15_data/zomato_sweetviz_compare.html')
print(f"Mumbai Mean Average Cost for Two: ₹{df_mumbai['Average_Cost_for_Two'].mean():.2f}")
print(f"Delhi Mean Average Cost for Two: ₹{df_delhi['Average_Cost_for_Two'].mean():.2f}")


Report session15_data/zomato_sweetviz_compare.html was generated.
Comparison Report saved to: session15_data/zomato_sweetviz_compare.html
Mumbai Mean Average Cost for Two: ₹1338.67
Delhi Mean Average Cost for Two: ₹784.00


**Key Differences Observed in Visual Report**:
1. **Average Cost for Two**: Mumbai restaurants have a significantly higher mean cost for two (₹1,338.67) compared to Delhi (₹784.00).
2. **Online Delivery**: Mumbai shows higher online delivery availability (88%) compared to Delhi (70%).

### Task 3: Myntra Product Listings - D-Tale Exploration & Filtering

**Objective**: Open Myntra product listings in D-Tale and filter products with price above ₹2000.

In [3]:
import pandas as pd
import dtale

# Load Myntra dataset
myntra_df = pd.read_csv('session15_data/myntra_products.csv')

# Filter products with Price > 2000
filtered_df = myntra_df[myntra_df['Price'] > 2000]

print(f'Total products: {len(myntra_df)}')
print(f'Number of products with Price > ₹2000: {len(filtered_df)}')
print('\nFiltered Sample:')
print(filtered_df[['Product_ID', 'Brand', 'Product_Name', 'Category', 'Price']].head())

# D-Tale Launch Syntax:
# d = dtale.show(myntra_df)


Total products: 200
Number of products with Price > ₹2000: 162

Filtered Sample:
   Product_ID     Brand    Product_Name    Category  Price
0        1001      Zara  Apparel Item 1     Topwear   4921
1        1002      Biba  Apparel Item 2  Bottomwear   4323
2        1003     Levis  Apparel Item 3     Dresses   2950
4        1005      Zara  Apparel Item 5    Footwear   8277
5        1006  FabIndia  Apparel Item 6    Footwear   4195


### Task 4: Flipkart Product Reviews - Report Interpretation & Data Cleaning

**Objective**: Identify two columns in the Flipkart product reviews dataset that require data cleaning or transformation before further analysis.

In [4]:
import pandas as pd

# Load Flipkart reviews dataset
flipkart_df = pd.read_csv('session15_data/flipkart_reviews.csv')

print('=== Dataset Summary ===')
flipkart_df.info()

print('\n1. Sample values of Helpful_Count:')
print(flipkart_df['Helpful_Count'].head(5))

print('\n2. Missing values in Review_Date:')
print(f"{flipkart_df['Review_Date'].isnull().sum()} missing rows out of {len(flipkart_df)}")


=== Dataset Summary ===
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   Review_ID       150 non-null    str  
 1   Product_Name    150 non-null    str  
 2   Rating          150 non-null    int64
 3   Review_Summary  150 non-null    str  
 4   Review_Text     150 non-null    str  
 5   User_Name       150 non-null    str  
 6   Review_Date     131 non-null    str  
 7   Helpful_Count   150 non-null    str  
dtypes: int64(1), str(7)
memory usage: 32.9 KB

1. Sample values of Helpful_Count:
0    34 people found this helpful
1    31 people found this helpful
2    35 people found this helpful
3    12 people found this helpful
4     5 people found this helpful
Name: Helpful_Count, dtype: str

2. Missing values in Review_Date:
19 missing rows out of 150


**Columns Requiring Transformation/Cleaning**:
1. **`Helpful_Count`**: Stored as text strings (`"12 people found this helpful"`). Needs numeric extraction (`str.extract(r'(\d+)')`) and casting to integer.
2. **`Review_Date`**: Contains missing values (`NaN`) and text formatted dates requiring `pd.to_datetime()` parsing and imputation.

### Task 5: Swiggy Food Orders - Sweetviz Summary Report

**Objective**: Generate a summary report for Swiggy food order CSV using Sweetviz.

In [6]:
import pandas as pd
import sweetviz as sv

# Load Swiggy dataset
swiggy_df = pd.read_csv('session15_data/swiggy_orders.csv')

# Generate Sweetviz Summary Report
swiggy_report = sv.analyze(swiggy_df)
swiggy_report.show_html('session15_data/swiggy_sweetviz_report.html', open_browser=False)

print('Swiggy Summary Report saved to: session15_data/swiggy_sweetviz_report.html')
print('\n=== Summary Statistics ===')
print(swiggy_df.describe())


Report session15_data/swiggy_sweetviz_report.html was generated.
Swiggy Summary Report saved to: session15_data/swiggy_sweetviz_report.html

=== Summary Statistics ===
            Price  Delivery_Time_Mins  Customer_Rating
count  100.000000          100.000000       100.000000
mean   476.170000           40.480000         4.207000
std    221.704647           11.323953         0.444326
min    127.000000           20.000000         3.500000
25%    302.750000           30.750000         3.800000
50%    468.500000           42.500000         4.250000
75%    679.000000           49.000000         4.500000
max    849.000000           59.000000         5.000000
